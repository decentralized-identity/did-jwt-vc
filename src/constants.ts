export const JWT_ALG = 'ES256K-R'
export const DID_FORMAT = /^did:([a-zA-Z0-9_]+):([:[a-zA-Z0-9_.-]+)(\/[^#]*)?(#.*)?$/
export const JWT_FORMAT = /^[A-Za-z0-9-_=]+\.[A-Za-z0-9-_=]+\.?[A-Za-z0-9-_.+/=]*$/
export const DEFAULT_CONTEXT = 'https://www.w3.org/2018/credentials/v1'
export const DEFAULT_TYPE = 'VerifiableCredential'
