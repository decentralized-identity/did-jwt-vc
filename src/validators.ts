import { DEFAULT_CONTEXT, DEFAULT_VC_TYPE, DEFAULT_VP_TYPE, JWT_FORMAT } from './types'
import { JwtCredentialSubject, DateType } from './types'
import { VerifiableCredential } from '.'
import { asArray } from './converters'
import { JWT_ERROR } from 'did-jwt'

/**
 * Error prefixes used for known verification failure cases related to the
 * {@link https://www.w3.org/TR/vc-data-model/ | Verifiable Credential data model }
 */
export const VC_ERROR = {
  /**
   * Thrown when the credential or presentation being verified does not conform to the data model defined by
   * {@link https://www.w3.org/TR/vc-data-model/ | the spec}
   */
  SCHEMA_ERROR: 'schema_error',

  /**
   * Thrown when the input is not a JWT string
   */
  FORMAT_ERROR: 'format_error',

  /**
   * Thrown when verifying a presentation where `challenge` and/or `domain` don't match the expected values.
   */
  AUTH_ERROR: 'auth_error',
}

/**
 * Known validation or verification error prefixes.
 */
export const VC_JWT_ERROR = { ...VC_ERROR, ...JWT_ERROR }

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isDateObject(input: any): input is Date {
  return input && !isNaN(input) && Object.prototype.toString.call(input) === '[object Date]'
}

export function validateJwtFormat(value: VerifiableCredential): void {
  if (typeof value === 'string' && !value.match(JWT_FORMAT)) {
    throw new TypeError(`${VC_ERROR.FORMAT_ERROR}: "${value}" is not a valid JWT format`)
  }
}

// The main scenario we want to guard against is having a timestamp in milliseconds
// instead of seconds (ex: from new Date().getTime()).
// We will check the number of digits and assume that any number with 12 or more
// digits is a millisecond timestamp.
// 10 digits max is 9999999999 -> 11/20/2286 @ 5:46pm (UTC)
// 11 digits max is 99999999999 -> 11/16/5138 @ 9:46am (UTC)
// 12 digits max is 999999999999 -> 09/27/33658 @ 1:46am (UTC)
export function validateTimestamp(value: number | DateType): void {
  if (typeof value === 'number') {
    if (!(Number.isInteger(value) && value < 100000000000)) {
      throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: "${value}" is not a unix timestamp in seconds`)
    }
  } else if (typeof value === 'string') {
    validateTimestamp(Math.floor(new Date(value).valueOf() / 1000))
  } else if (!isDateObject(value)) {
    throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: "${value}" is not a valid time`)
  }
}

export function validateContext(value: string | string[]): void {
  const input = asArray(value)
  if (input.length < 1 || input.indexOf(DEFAULT_CONTEXT) === -1) {
    throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: @context is missing default context "${DEFAULT_CONTEXT}"`)
  }
}

export function validateVcType(value: string | string[]): void {
  const input = asArray(value)
  if (input.length < 1 || input.indexOf(DEFAULT_VC_TYPE) === -1) {
    throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: type is missing default "${DEFAULT_VC_TYPE}"`)
  }
}

export function validateVpType(value: string | string[]): void {
  const input = asArray(value)
  if (input.length < 1 || input.indexOf(DEFAULT_VP_TYPE) === -1) {
    throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: type is missing default "${DEFAULT_VP_TYPE}"`)
  }
}

export function validateCredentialSubject(value: JwtCredentialSubject): void {
  if (Object.keys(value).length === 0) {
    throw new TypeError(`${VC_ERROR.SCHEMA_ERROR}: credentialSubject must not be empty`)
  }
}
