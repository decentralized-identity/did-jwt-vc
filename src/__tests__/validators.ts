import * as validators from '../validators'
import {
  DID_A,
  INVALID_DID,
  EXTRA_CONTEXT_A,
  EXTRA_CONTEXT_B,
  EXTRA_TYPE_A,
  EXTRA_TYPE_B,
  VC_JWT
} from '.'
import { DEFAULT_CONTEXT, DEFAULT_TYPE } from '../constants'

describe('validators', () => {
  describe('validateDidFormat', () => {
    it('does not throw if the value is a valid did format', () => {
      expect(() => validators.validateDidFormat(DID_A)).not.toThrow()
    })
    it('throws a TypeError if the value is not a valid did format', () => {
      expect(() => validators.validateDidFormat(INVALID_DID)).toThrow(TypeError)
    })
  })

  describe('validateTimestamp', () => {
    it('does not throw if the value is a valid unix timestamp in seconds', () => {
      expect(() =>
        validators.validateTimestamp(Math.floor(new Date().getTime() / 1000))
      ).not.toThrow()
    })
    it('throws a TypeError if the value is a millisecond timestamp', () => {
      expect(() => validators.validateTimestamp(new Date().getTime())).toThrow(
        TypeError
      )
    })
    it('throws a TypeError if the value is not an integer', () => {
      expect(() =>
        validators.validateTimestamp(new Date().getTime() / 1000)
      ).toThrow(TypeError)
    })
  })

  describe('validateContext', () => {
    it('does not throw if the value contains only the default context', () => {
      expect(() => validators.validateContext([DEFAULT_CONTEXT])).not.toThrow()
    })
    it('does not throw if the value contains the default context and some user-defined ones', () => {
      expect(() =>
        validators.validateContext([
          DEFAULT_CONTEXT,
          EXTRA_CONTEXT_A,
          EXTRA_CONTEXT_B
        ])
      ).not.toThrow()
    })
    it('throws a TypeError the value contains no contexts', () => {
      expect(() => validators.validateContext([])).toThrow(TypeError)
    })
    it('throws a TypeError the value is missing the default context', () => {
      expect(() =>
        validators.validateContext([EXTRA_CONTEXT_A, EXTRA_CONTEXT_B])
      ).toThrow(TypeError)
    })
  })

  describe('validateType', () => {
    it('does not throw if the value contains only the default type', () => {
      expect(() => validators.validateType([DEFAULT_TYPE])).not.toThrow()
    })
    it('does not throw if the value contains the default type and some user-defined ones', () => {
      expect(() =>
        validators.validateType([DEFAULT_TYPE, EXTRA_TYPE_A, EXTRA_TYPE_B])
      ).not.toThrow()
    })
    it('throws a TypeError the value contains no types', () => {
      expect(() => validators.validateType([])).toThrow(TypeError)
    })
    it('throws a TypeError the value is missing the default type', () => {
      expect(() =>
        validators.validateType([EXTRA_TYPE_A, EXTRA_TYPE_B])
      ).toThrow(TypeError)
    })
  })

  describe('validateJwtFormat', () => {
    it('does not throw if the value is a valid JWT format', () => {
      expect(() => validators.validateJwtFormat(VC_JWT)).not.toThrow()
    })
    it('throws a TypeError if the value is not a valid JWT format', () => {
      expect(() => validators.validateJwtFormat('not a jwt')).toThrow(TypeError)
    })
  })

  describe('validateCredentialSubject', () => {
    it('does not throw if the value is an object with at least one attribute', () => {
      expect(() =>
        validators.validateCredentialSubject({ name: 'test' })
      ).not.toThrow()
    })
    it('throws a TypeError if the value is an object with no attributes', () => {
      expect(() => validators.validateCredentialSubject({})).toThrow(TypeError)
    })
  })
})
