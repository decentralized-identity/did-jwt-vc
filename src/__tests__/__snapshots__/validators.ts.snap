// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPresentation creates a valid Presentation JWT with extra optional fields 1`] = `
Object {
  "extra": 42,
  "iss": "did:ethr:0xf1232f840f3ad7d23fcdaa84d6c66dac24efb198",
  "vp": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      "https://www.w3.org/2018/credentials/examples/v1",
    ],
    "type": Array [
      "VerifiableCredential",
    ],
    "verifiableCredential": Array [
      "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NkstUiJ9.eyJpYXQiOjE1NjM4MjQ4MDksInN1YiI6ImRpZDpldGhyOjB4MTIzNDU2NzgiLCJuYmYiOjE1NjI5NTAyODI4MDEsInZjIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIiwiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjEiXSwidHlwZSI6WyJWZXJpZmlhYmxlQ3JlZGVudGlhbCIsIlVuaXZlcnNpdHlEZWdyZWVDcmVkZW50aWFsIl0sImNyZWRlbnRpYWxTdWJqZWN0Ijp7ImRlZ3JlZSI6eyJ0eXBlIjoiQmFjaGVsb3JEZWdyZWUiLCJuYW1lIjoiQmFjY2FsYXVyw6lhdCBlbiBtdXNpcXVlcyBudW3DqXJpcXVlcyJ9fX0sImlzcyI6ImRpZDpldGhyOjB4ZjEyMzJmODQwZjNhZDdkMjNmY2RhYTg0ZDZjNjZkYWMyNGVmYjE5OCJ9.uYSRgDNmZnz0k5rORCBIIzEahVask5eQ2PFZI2_JAatvrpZ2t_3iTvPmBy6Kzt2W20fw5jUJ7GoZXJqoba4UVQA",
    ],
  },
}
`;

exports[`createPresentation creates a valid Presentation JWT with required fields 1`] = `
Object {
  "iss": "did:ethr:0xf1232f840f3ad7d23fcdaa84d6c66dac24efb198",
  "vp": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      "https://www.w3.org/2018/credentials/examples/v1",
    ],
    "type": Array [
      "VerifiableCredential",
    ],
    "verifiableCredential": Array [
      "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NkstUiJ9.eyJpYXQiOjE1NjM4MjQ4MDksInN1YiI6ImRpZDpldGhyOjB4MTIzNDU2NzgiLCJuYmYiOjE1NjI5NTAyODI4MDEsInZjIjp7IkBjb250ZXh0IjpbImh0dHBzOi8vd3d3LnczLm9yZy8yMDE4L2NyZWRlbnRpYWxzL3YxIiwiaHR0cHM6Ly93d3cudzMub3JnLzIwMTgvY3JlZGVudGlhbHMvZXhhbXBsZXMvdjEiXSwidHlwZSI6WyJWZXJpZmlhYmxlQ3JlZGVudGlhbCIsIlVuaXZlcnNpdHlEZWdyZWVDcmVkZW50aWFsIl0sImNyZWRlbnRpYWxTdWJqZWN0Ijp7ImRlZ3JlZSI6eyJ0eXBlIjoiQmFjaGVsb3JEZWdyZWUiLCJuYW1lIjoiQmFjY2FsYXVyw6lhdCBlbiBtdXNpcXVlcyBudW3DqXJpcXVlcyJ9fX0sImlzcyI6ImRpZDpldGhyOjB4ZjEyMzJmODQwZjNhZDdkMjNmY2RhYTg0ZDZjNjZkYWMyNGVmYjE5OCJ9.uYSRgDNmZnz0k5rORCBIIzEahVask5eQ2PFZI2_JAatvrpZ2t_3iTvPmBy6Kzt2W20fw5jUJ7GoZXJqoba4UVQA",
    ],
  },
}
`;

exports[`createVerifiableCredential creates a valid Verifiable Credential JWT with extra optional fields 1`] = `
Object {
  "extra": 42,
  "iss": "did:ethr:0xf1232f840f3ad7d23fcdaa84d6c66dac24efb198",
  "nbf": 1562950282,
  "sub": "did:ethr:0x435df3eda57154cf8cf7926079881f2912f54db4",
  "vc": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      "https://www.w3.org/2018/credentials/examples/v1",
    ],
    "credentialSubject": Object {
      "degree": Object {
        "name": "Baccalauréat en musiques numériques",
        "type": "BachelorDegree",
      },
    },
    "type": Array [
      "VerifiableCredential",
      "UniversityDegreeCredential",
    ],
  },
}
`;

exports[`createVerifiableCredential creates a valid Verifiable Credential JWT with required fields 1`] = `
Object {
  "iss": "did:ethr:0xf1232f840f3ad7d23fcdaa84d6c66dac24efb198",
  "nbf": 1562950282,
  "sub": "did:ethr:0x435df3eda57154cf8cf7926079881f2912f54db4",
  "vc": Object {
    "@context": Array [
      "https://www.w3.org/2018/credentials/v1",
      "https://www.w3.org/2018/credentials/examples/v1",
    ],
    "credentialSubject": Object {
      "degree": Object {
        "name": "Baccalauréat en musiques numériques",
        "type": "BachelorDegree",
      },
    },
    "type": Array [
      "VerifiableCredential",
      "UniversityDegreeCredential",
    ],
  },
}
`;
